hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileCreateHandler.java

```diff
@@ -49,7 +85,76 @@ public void preHandle(@NonNull final PreHandleContext context) {
      *
      * @throws NullPointerException if one of the arguments is {@code null}
      */
-    public void handle() {
-        throw new UnsupportedOperationException("Not implemented");
+    public void handle(
+            @NonNull final HandleContext handleContext,
+            @NonNull final FileCreateTransactionBody fileCreateTransactionBody,
+            @NonNull final FileServiceConfig fileServiceConfig,
+            @NonNull final CreateFileRecordBuilder recordBuilder,
+            @NonNull final WritableFileStoreImpl fileStore) {
+
+        final var builder = new File.Builder();
+
+        if (fileCreateTransactionBody.hasKeys()) {
+            builder.keys(fileCreateTransactionBody.keys());
+        }
+
+        /* Validate if the current file can be created */
+        if (fileStore.sizeOfState() >= fileServiceConfig.maxNum()) {
+            throw new HandleException(MAX_ENTITIES_IN_PRICE_REGIME_HAVE_BEEN_CREATED);
+        }
+
+        final var entityExpiryMeta = new ExpiryMeta(
+                fileCreateTransactionBody.hasExpirationTime()
+                        ? fileCreateTransactionBody.expirationTimeOrThrow().seconds()
+                        : NA,
+                NA,
+                // Shard and realm will be ignored if num is NA
+                NA,
+                NA,
+                NA);
+
+        try {
+            final var effectiveExpiryMeta =
+                    handleContext.expiryValidator().resolveCreationAttempt(false, entityExpiryMeta);
+            builder.expirationTime(effectiveExpiryMeta.expiry());
+
+            handleContext.attributeValidator().validateMemo(fileCreateTransactionBody.memo());
+            builder.memo(fileCreateTransactionBody.memo());
+
+            builder.keys(fileCreateTransactionBody.keys());
+            builder.fileNumber(handleContext.newEntityNumSupplier().getAsLong());
+            validateContent(fileCreateTransactionBody, fileServiceConfig);
+            builder.contents(fileCreateTransactionBody.contents());
+
+            final var file = builder.build();
+            fileStore.put(file);
+
+            recordBuilder.setCreatedFile(file.fileNumber());
+        } catch (final HandleException e) {
+            if (e.getStatus() == INVALID_EXPIRATION_TIME) {
+                // Since for some reason CreateTransactionBody does not have an expiration time,
+                // it makes more sense to propagate AUTORENEW_DURATION_NOT_IN_RANGE
+                var handleException = new HandleException(AUTORENEW_DURATION_NOT_IN_RANGE);
```

Comment:
Probably a separate PR, but we should add a constructor to `HandleException` that accepts a cause.

Reply:
I fixed it already.